### PR TITLE
[TEP-0112] Replace volumes with Workspaces [Proposal]

### PR DESCRIPTION
**Pull request by lbernick**
This TEP proposes removing volumes from the Task spec in the v1 API, as well as removing
volumeMounts and volumeDevices for the container-like fields (Step, StepTemplate, and Sidecar).

/kind tep

### CODE REVIEW

**Comment by lbernick** [comment]
Addresses https://github.com/tektoncd/pipeline/issues/2058

@vdemeester @skaegi PTAL

**Comment by lbernick** [comment]
/hold

need to do a bit more digging--will revisit the proposal to support hostPath
@wlynch relevant for you as well

File: teps/0111-remove-volumes-from-task-spec.md
Comment (pierretasci) [inline teps/0111-remove-volumes-from-task-spec.md:143]: We have another use case not already covered here and not supported by workspaces (nor could they easily be) which is dynamic volumes. Sometimes, it is desired/opportune to create a volume (say backed by a configmap or secret) dynamically in a pipeline and then mount it into an already named volume mount in a task. This is not possible with workspaces because the workspace bindings need to be defined up front.

**Comment by lbernick** [comment]
I chatted with @pierretasci yesterday about the "dynamic volumes" use case, and added it here as an existing use case that workspaces don't support. I need to think more about what we should do about this, but for now I added "dynamic workspaces" as an optional workspace enhancement. We are both in agreement that this is the right long-term direction for volumes/workspaces, but want to avoid removing support for an existing use case even if it isn't an "intended" use of the API.

I've also added more detail on the options for different volume types we might want to support in workspaces, including embedding the k8s volumesource in workspacebinding. For now I think the existing workspace binding options are sufficient for the use cases we're targeting.

File: teps/0111-replace-volumes-with-workspaces.md
Comment (afrittoli) [inline teps/0111-replace-volumes-with-workspaces.md:119]: What would be the alternative here if we removed workspaces?
I agree that using volumes directly instead of workspaces is not a best practice, but unless we have an alternative here I wonder if we could address this by documentation, i.e. make sure that volumes are not used in any example / catalog task unless were strictly required. We could add a special page / section to document this feature as something we do not recommend to use unless in specific documented cases and specify security risks.

**Comment by lbernick** [comment]
@pritidesai @afrittoli @pierretasci thanks for the feedback so far!
I think there are a few too many open questions so I've marked this as "proposed", removed the "proposal" section, and put the ideas being floated under "alternatives" (can always add more). Let's align on the use cases we're addressing and then we can figure out what to do about it.

File: teps/0111-replace-volumes-with-workspaces.md
Comment (lbernick) [reply teps/0111-replace-volumes-with-workspaces.md:118]: I've added this as an option to "alternatives". Right now, I don't think we would have an alternative for this using workspaces.

File: teps/0111-replace-volumes-with-workspaces.md
Comment (lbernick) [reply teps/0111-replace-volumes-with-workspaces.md:190]: nice idea, I added this to alternatives as well and put pros + cons for both solutions.

File: teps/0111-replace-volumes-with-workspaces.md
Comment (dibyom) [reply teps/0111-replace-volumes-with-workspaces.md:202]: Ah, looks like this is the next alternative :)

File: teps/0111-replace-volumes-with-workspaces.md
Comment (dibyom) [inline teps/0111-replace-volumes-with-workspaces.md:325]: Looking at both the internal and this hostPath use case, it seems like there are aa few use cases where the task authors need to provide the volume/workspace type they need and workspaces today only allow taskrun authors to supply these

File: teps/0111-replace-volumes-with-workspaces.md
Comment (Tomcli) [inline teps/0111-replace-volumes-with-workspaces.md:348]: If we are going to remove volumemount and volumes for tasks, will there be a document on how to migrate podTemplate.volumes with workspaces?

File: teps/0111-remove-volumes-from-task-spec.md
Comment (lbernick) [reply teps/0111-remove-volumes-from-task-spec.md]: resolving since this is now in "proposed" state with proposal removed

File: teps/0111-replace-volumes-with-workspaces.md
Comment (lbernick) [reply teps/0111-replace-volumes-with-workspaces.md:46]: done! moved under "proposal"

File: teps/0111-replace-volumes-with-workspaces.md
Comment (lbernick) [reply teps/0111-replace-volumes-with-workspaces.md:325]: agreed! I think this is covered under the "task-internal" workspaces option

File: teps/0111-replace-volumes-with-workspaces.md
Comment (lbernick) [reply teps/0111-replace-volumes-with-workspaces.md:348]: Definitely! I added a section about this under "design considerations".

File: teps/0111-remove-volumes-from-task-spec.md
Comment (lbernick) [reply teps/0111-remove-volumes-from-task-spec.md]: Thanks Pierre! I added this as a use case that workspaces don't currently support, and modified the TEP to specify that before it's merged as implementable, we should have a plan for workspaces support for this use case. PTAL!

**Review by dibyom — APPROVED** [review]

**Pull request merged**
